src/nvim/eval/decode.c

```diff
@@ -0,0 +1,945 @@
+#include <stddef.h>
+
+#include <msgpack.h>
+
+#include "nvim/eval_defs.h"
+#include "nvim/eval.h"
+#include "nvim/eval/encode.h"
+#include "nvim/ascii.h"
+#include "nvim/message.h"
+#include "nvim/charset.h"  // vim_str2nr
+#include "nvim/lib/kvec.h"
+#include "nvim/vim.h"  // OK, FAIL
+
+/// Helper structure for container_struct
+typedef struct {
+  size_t stack_index;   ///< Index of current container in stack.
+  list_T *special_val;  ///< _VAL key contents for special maps.
+                        ///< When container is not a special dictionary it is
+                        ///< NULL.
+  const char *s;        ///< Location where container starts.
+  typval_T container;   ///< Container. Either VAR_LIST, VAR_DICT or VAR_LIST
+                        ///< which is _VAL from special dictionary.
+} ContainerStackItem;
+
+/// Helper structure for values struct
+typedef struct {
+  bool is_special_string;  ///< Indicates that current value is a special
+                           ///< dictionary with string.
+  bool didcomma;           ///< True if previous token was comma.
+  bool didcolon;           ///< True if previous token was colon.
+  typval_T val;            ///< Actual value.
+} ValuesStackItem;
+
+/// Vector containing values not yet saved in any container
+typedef kvec_t(ValuesStackItem) ValuesStack;
+
+/// Vector containing containers, each next container is located inside previous
+typedef kvec_t(ContainerStackItem) ContainerStack;
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "eval/decode.c.generated.h"
+#endif
+
+/// Create special dictionary
+///
+/// @param[out]  rettv  Location where created dictionary will be saved.
+/// @param[in]  type  Type of the dictionary.
+/// @param[in]  val  Value associated with the _VAL key.
+static inline void create_special_dict(typval_T *const rettv,
+                                       const MessagePackType type,
+                                       typval_T val)
+  FUNC_ATTR_NONNULL_ALL
+{
+  dict_T *const dict = dict_alloc();
+  dictitem_T *const type_di = dictitem_alloc((char_u *) "_TYPE");
+  type_di->di_tv.v_type = VAR_LIST;
+  type_di->di_tv.v_lock = 0;
+  type_di->di_tv.vval.v_list = (list_T *) eval_msgpack_type_lists[type];
+  type_di->di_tv.vval.v_list->lv_refcount++;
+  dict_add(dict, type_di);
+  dictitem_T *const val_di = dictitem_alloc((char_u *) "_VAL");
+  val_di->di_tv = val;
+  dict_add(dict, val_di);
+  dict->dv_refcount++;
+  *rettv = (typval_T) {
+    .v_type = VAR_DICT,
+    .v_lock = VAR_UNLOCKED,
+    .vval = { .v_dict = dict },
+  };
+}
+
+#define DICT_LEN(dict) (dict)->dv_hashtab.ht_used
+
+/// Helper function used for working with stack vectors used by JSON decoder
+///
+/// @param[in]  obj  New object.
```

Comment:
Missing `out`:
`obj` is a copy, but it contains pointers that could be freed (`clear_tv(&obj.val)`).